png: add support for 16-bit PNG files.

krasin opened the pull request:
Note: libpng.load now returns two arguments. The first argument is the result tensor,
and the second argument is the color bitness. It will be either 8 or 16. For images
with smaller color bit depth (for example, 4), the value is upgraded to 8 bits.

See also #74 and #75.

soumith commented on init.lua:
```diff
@@ -118,8 +118,9 @@ local function decompress(tensor, depth, tensortype)
 rawset(image, 'decompress', decompress)
 
-local function processPNG(img, depth, tensortype)
+local function processPNG(img, depth, bit_depth, tensortype)
```
make bit_depth to be the last argument, so as to not break existing scripts

soumith replied:
oh nevermind, no one uses this directly.

soumith commented:
this looks pretty good to me overall

soumith commented:
have to tested this on both 8-bit and 16-bit stuff? can you add some simple tests (even just to display a couple of images for sanity checks) to the test folder

krasin commented:
@soumith thanks for your comments.

re: tests. I would like to add PNG tests to [torch/image/test](https://github.com/torch/image/tree/master/test) dir, just like you have for PPM, PGM and JPG.

One thing that I failed to figure out myself, is how to run these tests. I don't see them hooked up in [CMakeLists.txt](https://github.com/torch/image/blob/master/CMakeLists.txt). Am I missing something?

soumith commented:
we just run them manually with the interpreter. For example:
th test_compress_jpg.lua

krasin commented:
@soumith I have made the first draft for test_png.lua, please take a look at the commit above.

I generate real PNG files of a very small size and load them with different options:

```
gray16-1x2.png: PNG image data, 1 x 2, 16-bit grayscale, non-interlaced
gray3x1.png:    PNG image data, 3 x 1, 8-bit grayscale, non-interlaced
```

I also plan to  cover color images, but I would like to hear your feedback on the test code first. Is there a better way to specify a Tensor literal?

soumith commented on test/test_png.lua:
```diff
@@ -0,0 +1,44 @@
+
+function test.LoadPNG()
+  -- Gray 8-bit PNG image with width = 3, height = 1
+  local gray8byte = torch.ByteTensor(1, 1, 3)
```
torch.ByteTensor({{{0,127,255}}})

krasin replied:
Thanks, done.

soumith commented on test/test_png.lua:
```diff
@@ -0,0 +1,44 @@
+  gray8byte[1][1][3] = 255
+  checkPNG('gray3x1.png', 1, 'byte', gray8byte)
+
+  local gray8double = torch.DoubleTensor(1, 1, 3)
```
torch.ByteTensor({{{0,127/255,1}}})

krasin replied:
Done.

soumith commented on test/test_png.lua:
```diff
@@ -0,0 +1,44 @@
+
+
+  -- Gray 16-bit PNG image with width=1, height = 2
+  local gray16byte = torch.ByteTensor(1, 2, 1)
```
torch.ByteTensor({{{0},{255}}})

krasin replied:
Done.

soumith commented:
yes, gave example in in-line comments.

krasin commented:
@soumith thank you!

I have added all the tests I wanted (8-bit / 16-bit, gray / RGB) and fixed a bug with 16-bit RGB images in the process of doing that.

Please, take another look!

krasin commented:
Hi @soumith,

is there anything I can do to improve this CL?
I would appreciate any feedback.

Thanks! :)

soumith commented:
no, this looks perfect, I'm sorry for the delayed response, sometimes things get lost in the haystack. thanks again for the PR.

krasin commented:
@soumith thanks!
